agent/rpc/subscribe/subscribe.go

```diff
@@ -0,0 +1,212 @@
+package subscribe
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/go-hclog"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/consul/state"
+	"github.com/hashicorp/consul/agent/consul/stream"
+	"github.com/hashicorp/consul/proto/pbservice"
+	"github.com/hashicorp/consul/proto/pbsubscribe"
+)
+
+// Server implements a StateChangeSubscriptionServer for accepting SubscribeRequests,
+// and sending events to the subscription topic.
+type Server struct {
+	Backend Backend
+	Logger  Logger
+}
+
+func NewServer(backend Backend, logger Logger) *Server {
+	return &Server{Backend: backend, Logger: logger}
+}
+
+type Logger interface {
+	Trace(msg string, args ...interface{})
+	With(args ...interface{}) hclog.Logger
+}
+
+var _ pbsubscribe.StateChangeSubscriptionServer = (*Server)(nil)
+
+type Backend interface {
+	ResolveToken(token string) (acl.Authorizer, error)
+	Forward(dc string, f func(*grpc.ClientConn) error) (handled bool, err error)
+	Subscribe(req *stream.SubscribeRequest) (*stream.Subscription, error)
+}
+
+func (h *Server) Subscribe(req *pbsubscribe.SubscribeRequest, serverStream pbsubscribe.StateChangeSubscription_SubscribeServer) error {
+	logger := h.newLoggerForRequest(req)
+	handled, err := h.Backend.Forward(req.Datacenter, forwardToDC(req, serverStream, logger))
+	if handled || err != nil {
+		return err
+	}
+
+	logger.Trace("new subscription")
+	defer logger.Trace("subscription closed")
+
+	// Resolve the token and create the ACL filter.
+	// TODO(streaming): handle token expiry gracefully...
```

Comment:
Do you mean a token expiring during the stream or being expired at this point in the code. Because if a token is actually expired right here ResolveToken should return some sort of ACL error.

In the other case of being valid now but expiring later you should periodically re-resolve the token. For the xDS server we have a ticker which fires every so often to force authorization of the stream to happen again and it seems like a similar change could be made here as a first step.

Reply:
The way it works right now (https://github.com/hashicorp/consul/blob/master/agent/consul/state/acl_events.go) is that if any of the ACLs change we build up a list of tokens that are impacted, create a special event that is handled by EventPublisher, and it closes any of the subscription which used those tokens, with a specific error.

This TODO came from the original diff, but I left it because I think the idea is that force closing all the impacted subscriptions is not "graceful". I don't think it's something we need to fix for the first release.